File: test/aidbox_sdk/snapshots/dotnet/hl7-fhir-r4-core/MedicationKnowledge.cs
using Aidbox.FHIR.Base;
using Aidbox.FHIR.Utils;

namespace Aidbox.FHIR.R4.Core;

public class MedicationKnowledge : DomainResource, IResource
{
    public string? PreparationInstruction { get; set; }
    public Base.Quantity? Amount { get; set; }
    public MedicationKnowledge_Monograph[]? Monograph { get; set; }
    public MedicationKnowledge_Regulatory[]? Regulatory { get; set; }
    public Base.CodeableConcept? DoseForm { get; set; }
    public Base.CodeableConcept[]? IntendedRoute { get; set; }
    public MedicationKnowledge_DrugCharacteristic[]? DrugCharacteristic { get; set; }
    public MedicationKnowledge_Packaging? Packaging { get; set; }
    public MedicationKnowledge_RelatedMedicationKnowledge[]? RelatedMedicationKnowledge { get; set; }
    public MedicationKnowledge_MedicineClassification[]? MedicineClassification { get; set; }
    public MedicationKnowledge_Kinetics[]? Kinetics { get; set; }
    public Base.ResourceReference[]? AssociatedMedication { get; set; }
    public MedicationKnowledge_Ingredient[]? Ingredient { get; set; }
    public MedicationKnowledge_MonitoringProgram[]? MonitoringProgram { get; set; }
    public Base.ResourceReference[]? Contraindication { get; set; }
    public string? Status { get; set; }
    public Base.CodeableConcept[]? ProductType { get; set; }
    public string[]? Synonym { get; set; }
    public Base.CodeableConcept? Code { get; set; }
    public MedicationKnowledge_AdministrationGuidelines[]? AdministrationGuidelines { get; set; }
    public Base.ResourceReference? Manufacturer { get; set; }
    public MedicationKnowledge_Cost[]? Cost { get; set; }

    public class MedicationKnowledge_Monograph : BackboneElement
    {
        public Base.CodeableConcept? Type { get; set; }
        public Base.ResourceReference? Source { get; set; }
    }

    public class MedicationKnowledge_Regulatory_Schedule : BackboneElement
    {
        public required Base.CodeableConcept Schedule { get; set; }
    }

    public class MedicationKnowledge_Regulatory_MaxDispense : BackboneElement
    {
        public string? Period { get; set; }
        public required Base.Quantity Quantity { get; set; }
    }

    public class MedicationKnowledge_Regulatory_Substitution : BackboneElement
    {
        public required Base.CodeableConcept Type { get; set; }
        public required bool Allowed { get; set; }
    }

    public class MedicationKnowledge_Regulatory : BackboneElement
    {
        public MedicationKnowledge_Regulatory_Schedule[]? Schedule { get; set; }
        public MedicationKnowledge_Regulatory_MaxDispense? MaxDispense { get; set; }
        public MedicationKnowledge_Regulatory_Substitution[]? Substitution { get; set; }
        public required Base.ResourceReference RegulatoryAuthority { get; set; }
    }

    public class MedicationKnowledge_DrugCharacteristic : BackboneElement
    {
        public Base.CodeableConcept? Type { get; set; }
        public object? Value    
        {
            get
            {
                if (ValueString is not null)
                {
                    return ValueString;
                }
        
                if (ValueQuantity is not null)
                {
                    return ValueQuantity;
                }
        
                if (ValueBase64Binary is not null)
                {
                    return ValueBase64Binary;
                }
        
                if (ValueCodeableConcept is not null)
                {
                    return ValueCodeableConcept;
                }
        
                return null;
            }
        
            set
            {
                if (value?.GetType() == typeof(string))
                {
                    ValueString = (string)value;return;
                }
        
                if (value?.GetType() == typeof(Base.Quantity))
                {
                    ValueQuantity = (Base.Quantity)value;return;
                }
        
                if (value?.GetType() == typeof(string))
                {
                    ValueBase64Binary = (string)value;return;
                }
        
                if (value?.GetType() == typeof(Base.CodeableConcept))
                {
                    ValueCodeableConcept = (Base.CodeableConcept)value;return;
                }
        
                throw new ArgumentException("Invalid type provided");
            }
        }
        public string? ValueString { get; set; }
        public Base.Quantity? ValueQuantity { get; set; }
        public string? ValueBase64Binary { get; set; }
        public Base.CodeableConcept? ValueCodeableConcept { get; set; }
    }

    public class MedicationKnowledge_Packaging : BackboneElement
    {
        public Base.CodeableConcept? Type { get; set; }
        public Base.Quantity? Quantity { get; set; }
    }

    public class MedicationKnowledge_RelatedMedicationKnowledge : BackboneElement
    {
        public required Base.CodeableConcept Type { get; set; }
        public required Base.ResourceReference[] Reference { get; set; }
    }

    public class MedicationKnowledge_MedicineClassification : BackboneElement
    {
        public required Base.CodeableConcept Type { get; set; }
        public Base.CodeableConcept[]? Classification { get; set; }
    }

    public class MedicationKnowledge_Kinetics : BackboneElement
    {
        public Base.Quantity[]? LethalDose50 { get; set; }
        public Base.Quantity[]? AreaUnderCurve { get; set; }
        public string? HalfLifePeriod { get; set; }
    }

    public class MedicationKnowledge_Ingredient : BackboneElement
    {
        public object? Item    
        {
            get
            {
                if (ItemReference is not null)
                {
                    return ItemReference;
                }
        
                if (ItemCodeableConcept is not null)
                {
                    return ItemCodeableConcept;
                }
        
                return null;
            }
        
            set
            {
                if (value?.GetType() == typeof(Base.ResourceReference))
                {
                    ItemReference = (Base.ResourceReference)value;return;
                }
        
                if (value?.GetType() == typeof(Base.CodeableConcept))
                {
                    ItemCodeableConcept = (Base.CodeableConcept)value;return;
                }
        
                throw new ArgumentException("Invalid type provided");
            }
        }
        public bool? IsActive { get; set; }
        public Base.Ratio? Strength { get; set; }
        public Base.ResourceReference? ItemReference { get; set; }
        public Base.CodeableConcept? ItemCodeableConcept { get; set; }
    }

    public class MedicationKnowledge_MonitoringProgram : BackboneElement
    {
        public string? Name { get; set; }
        public Base.CodeableConcept? Type { get; set; }
    }

    public class MedicationKnowledge_AdministrationGuidelines_Dosage : BackboneElement
    {
        public required Base.CodeableConcept Type { get; set; }
        public required string[] Dosage { get; set; }
    }

    public class MedicationKnowledge_AdministrationGuidelines_PatientCharacteristics : BackboneElement
    {
        public string[]? Value { get; set; }
        public object? Characteristic    
        {
            get
            {
                if (CharacteristicQuantity is not null)
                {
                    return CharacteristicQuantity;
                }
        
                if (CharacteristicCodeableConcept is not null)
                {
                    return CharacteristicCodeableConcept;
                }
        
                return null;
            }
        
            set
            {
                if (value?.GetType() == typeof(Base.Quantity))
                {
                    CharacteristicQuantity = (Base.Quantity)value;return;
                }
        
                if (value?.GetType() == typeof(Base.CodeableConcept))
                {
                    CharacteristicCodeableConcept = (Base.CodeableConcept)value;return;
                }
        
                throw new ArgumentException("Invalid type provided");
            }
        }
        public Base.Quantity? CharacteristicQuantity { get; set; }
        public Base.CodeableConcept? CharacteristicCodeableConcept { get; set; }
    }

    public class MedicationKnowledge_AdministrationGuidelines : BackboneElement
    {
        public MedicationKnowledge_AdministrationGuidelines_Dosage[]? Dosage { get; set; }
        public object? Indication    
        {
            get
            {
                if (IndicationReference is not null)
                {
                    return IndicationReference;
                }
        
                if (IndicationCodeableConcept is not null)
                {
                    return IndicationCodeableConcept;
                }
        
                return null;
            }
        
            set
            {
                if (value?.GetType() == typeof(Base.ResourceReference))
                {
                    IndicationReference = (Base.ResourceReference)value;return;
                }
        
                if (value?.GetType() == typeof(Base.CodeableConcept))
                {
                    IndicationCodeableConcept = (Base.CodeableConcept)value;return;
                }
        
                throw new ArgumentException("Invalid type provided");
            }
        }
        public Base.ResourceReference? IndicationReference { get; set; }
        public MedicationKnowledge_AdministrationGuidelines_PatientCharacteristics[]? PatientCharacteristics { get; set; }
        public Base.CodeableConcept? IndicationCodeableConcept { get; set; }
    }

    public class MedicationKnowledge_Cost : BackboneElement
    {
        public required Base.Money Cost { get; set; }
        public required Base.CodeableConcept Type { get; set; }
        public string? Source { get; set; }
    }
}

File: test/aidbox_sdk/snapshots/python/hl7-fhir-r4-core/MedicationKnowledge.py
from typing import Optional, List
from pydantic import *
from ..base import *

class MedicationKnowledge_Monograph(BackboneElement):
    type: Optional[CodeableConcept] = None
    source: Optional[Reference] = None

class MedicationKnowledge_Regulatory_Schedule(BackboneElement):
    schedule: CodeableConcept

class MedicationKnowledge_Regulatory_MaxDispense(BackboneElement):
    period: Optional[Duration] = None
    quantity: Quantity

class MedicationKnowledge_Regulatory_Substitution(BackboneElement):
    type: CodeableConcept
    allowed: bool

class MedicationKnowledge_Regulatory(BackboneElement):
    schedule: Optional[List[MedicationKnowledge_Regulatory_Schedule]] = None
    max_dispense: Optional[MedicationKnowledge_Regulatory_MaxDispense] = None
    substitution: Optional[List[MedicationKnowledge_Regulatory_Substitution]] = None
    regulatory_authority: Reference

class MedicationKnowledge_DrugCharacteristic(BackboneElement):
    type: Optional[CodeableConcept] = None
    value_string: Optional[str] = None
    value_quantity: Optional[Quantity] = None
    value_base64_binary: Optional[str] = None
    value_codeable_concept: Optional[CodeableConcept] = None

class MedicationKnowledge_Packaging(BackboneElement):
    type: Optional[CodeableConcept] = None
    quantity: Optional[Quantity] = None

class MedicationKnowledge_RelatedMedicationKnowledge(BackboneElement):
    type: CodeableConcept
    reference: list[Reference] = []

class MedicationKnowledge_MedicineClassification(BackboneElement):
    type: CodeableConcept
    classification: Optional[List[CodeableConcept]] = None

class MedicationKnowledge_Kinetics(BackboneElement):
    lethal_dose50: Optional[List[Quantity]] = None
    area_under_curve: Optional[List[Quantity]] = None
    half_life_period: Optional[Duration] = None

class MedicationKnowledge_Ingredient(BackboneElement):
    is_active: Optional[bool] = None
    strength: Optional[Ratio] = None
    item_reference: Optional[Reference] = None
    item_codeable_concept: Optional[CodeableConcept] = None

class MedicationKnowledge_MonitoringProgram(BackboneElement):
    name: Optional[str] = None
    type: Optional[CodeableConcept] = None

class MedicationKnowledge_AdministrationGuidelines_Dosage(BackboneElement):
    type: CodeableConcept
    dosage: list[Dosage] = []

class MedicationKnowledge_AdministrationGuidelines_PatientCharacteristics(BackboneElement):
    value: Optional[List[str]] = None
    characteristic_quantity: Optional[Quantity] = None
    characteristic_codeable_concept: Optional[CodeableConcept] = None

class MedicationKnowledge_AdministrationGuidelines(BackboneElement):
    dosage: Optional[List[MedicationKnowledge_AdministrationGuidelines_Dosage]] = None
    indication_reference: Optional[Reference] = None
    patient_characteristics: Optional[List[MedicationKnowledge_AdministrationGuidelines_PatientCharacteristics]] = None
    indication_codeable_concept: Optional[CodeableConcept] = None

class MedicationKnowledge_Cost(BackboneElement):
    cost: Money
    type: CodeableConcept
    source: Optional[str] = None

class MedicationKnowledge(DomainResource):
    preparation_instruction: Optional[str] = None
    amount: Optional[Quantity] = None
    monograph: Optional[List[MedicationKnowledge_Monograph]] = None
    regulatory: Optional[List[MedicationKnowledge_Regulatory]] = None
    dose_form: Optional[CodeableConcept] = None
    intended_route: Optional[List[CodeableConcept]] = None
    drug_characteristic: Optional[List[MedicationKnowledge_DrugCharacteristic]] = None
    packaging: Optional[MedicationKnowledge_Packaging] = None
    related_medication_knowledge: Optional[List[MedicationKnowledge_RelatedMedicationKnowledge]] = None
    medicine_classification: Optional[List[MedicationKnowledge_MedicineClassification]] = None
    kinetics: Optional[List[MedicationKnowledge_Kinetics]] = None
    associated_medication: Optional[List[Reference]] = None
    ingredient: Optional[List[MedicationKnowledge_Ingredient]] = None
    monitoring_program: Optional[List[MedicationKnowledge_MonitoringProgram]] = None
    contraindication: Optional[List[Reference]] = None
    status: Optional[str] = None
    product_type: Optional[List[CodeableConcept]] = None
    synonym: Optional[List[str]] = None
    code: Optional[CodeableConcept] = None
    administration_guidelines: Optional[List[MedicationKnowledge_AdministrationGuidelines]] = None
    manufacturer: Optional[Reference] = None
    cost: Optional[List[MedicationKnowledge_Cost]] = None